Organise unit tests to use pytest fixtures

SeanBryan51 opened the pull request:
Unit tests have an excessive amount of repeated test setup code, making unit tests harder to maintain and read. Using pytest fixtures to setup each test would remove needless code repetition and also prevents the possibility of state leaking into each test case.

Organise the tests so that for each function `func`, we have a test class `TestFunc` with each method of `TestFunc` containing a single test (success or failure case).

Use pytest's `parametrize()` feature for testing different levels of verbosity.

Remove usage of `mock_cwd` fixture unless it is required for a test to pass.

Fixes #163

SeanBryan51 commented:
I have reorganised a few tests files just to see what it looks like. @ccarouge @bschroeter it would be great if I can get your thoughts before I push on further.

ccarouge commented on tests/test_comparison.py:
```diff
@@ -3,86 +3,98 @@
-    )
+    @pytest.fixture
+    def bitwise_cmp_dir(self, mock_cwd):
+        _bitwise_cmp_dir = mock_cwd / internal.FLUXSITE_BITWISE_CMP_DIR
```
With the `mock_cwd` fixture, it means you can use a relative path here (internal.FLUXSITE_BITWISE_CMP_DIR only). It might lighten up the tests. Might be applicable elsewhere.

ccarouge commented on tests/test_comparison.py:
```diff
@@ -3,86 +3,98 @@
+        comparison_task.run()
+        assert f"nccmp -df {file_a} {file_b}" in mock_subprocess_handler.commands
+
+    def test_default_standard_output(self, comparison_task, files):
```
I wonder if we could instead use the test [parameterization](https://docs.pytest.org/en/7.1.x/how-to/parametrize.html) to write one output test that runs on all output cases. But this might wait for after the reorganisation?

SeanBryan51 replied:
It should be possible to use parameterisation for testing different verbosity levels. I'll look into it

ccarouge reviewed (commented):
I think it looks better than before.

But considering the amount of changes, do we have to do all the tests in the same PR? It might make it very very hard to review. Could we do one module per PR or only a couple at a time?

SeanBryan51 commented:
> do we have to do all the tests in the same PR? It might make it very very hard to review. Could we do one module per PR or only a couple at a time?

I think I prefer we do the merge into main as a single PR. But we can treat this branch as 'main' for this issue and create other branches that implement small changes and have those merged into this branch?

ccarouge commented:
> But we can treat this branch as 'main' for this issue and create other branches that implement small changes and have those merged into this branch?

Yes, that would be better!

SeanBryan51 commented:
@ccarouge I've rebased this branch off main so that we have the recently merged changes (e.g. linting with ruff). I've added the changes you requested for using pytest's `parametrize()` and usage of `mock_cwd`.

ccarouge commented:
@SeanBryan51 It seems you asked my review again on this meta-PR. Is that correct?

SeanBryan51 commented:
@ccarouge Yes, just checking I've addressed your comments in: https://github.com/CABLE-LSM/benchcab/pull/172/commits/19eff9e43145c0266ee074bbe1e3b448e5a6415b
